src/coreclr/jit/target.h

```diff
@@ -208,8 +208,156 @@ enum _regMask_enum : unsigned
 // In any case, we believe that is OK to freely cast between these types; no information will
 // be lost.
 
-#if defined(TARGET_AMD64) || defined(TARGET_ARMARCH) || defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
+typedef _regNumber_enum regNumber;
+typedef unsigned char   regNumberSmall;
+
+#if defined(TARGET_AMD64) || defined(TARGET_ARM) || defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
 typedef unsigned __int64 regMaskTP;
+#elif defined(TARGET_ARM64)
+typedef unsigned __int64 regMaskSmall;
+struct regMaskTP
+{
+    unsigned __int64 low;
+
+    regMaskTP(unsigned __int64 regMask)
+        : low(regMask)
+    {
+    }
+
+    regMaskTP()
+    {
+    }
+
+    FORCEINLINE explicit operator bool() const
+    {
+        return low != 0;
+    }
+
+    FORCEINLINE explicit operator regMaskSmall() const
+    {
+        return (regMaskSmall)low;
+    }
+
+    FORCEINLINE explicit operator unsigned int() const
+    {
+        return (unsigned int)low;
+    }
+
+    FORCEINLINE unsigned __int64 getLow() const
+    {
+        return low;
+    }
+};
+
+static uint32_t PopCount(regMaskTP value)
+{
+    return BitOperations::PopCount(value.getLow());
+}
+
+static uint32_t BitScanForward(regMaskTP mask)
+{
+    return BitOperations::BitScanForward(mask.low);
+}
+
+static regMaskTP operator^(regMaskTP first, regMaskTP second)
+{
+    regMaskTP result(first.low ^ second.getLow());
+    return result;
+}
+
+static regMaskTP operator&(regMaskTP first, regMaskTP second)
+{
+    regMaskTP result(first.low & second.getLow());
+    return result;
+}
+
+static regMaskTP operator|(regMaskTP first, regMaskTP second)
+{
+    regMaskTP result(first.low | second.getLow());
+    return result;
+}
+
+static regMaskTP operator<<(regMaskTP first, const int b)
+{
+    regMaskTP result(first.low << b);
+    return result;
+}
+
+static regMaskTP operator>>(regMaskTP first, const int b)
+{
+    regMaskTP result(first.low >> b);
+    return result;
+}
+
+static regMaskTP& operator>>=(regMaskTP& first, const int b)
+{
+    first = first >> b;
+    return first;
+}
+
+static regMaskTP& operator|=(regMaskTP& first, regMaskTP second)
+{
+    first.low |= second.low;
+    return first;
+}
+
+static regMaskTP& operator^=(regMaskTP& first, regMaskTP second)
+{
+    first.low ^= second.low;
+    return first;
+}
+
+static regMaskSmall operator^=(regMaskSmall& first, regMaskTP second)
+{
+    first ^= second.low;
+    return first;
+}
```

Comment:
For my education, do you know what the point of `regMaskSmall` actually is?

Reply:
From the comments, I do not fully understand too and I am planning to audit it once this work is done. 

```
// regMaskSmall is large enough to represent the entire set of registers.
// If regMaskSmall is smaller than a "natural" integer type, regMaskTP is wider, based
// on a belief by the original authors of the JIT that in some situations it is more
// efficient to have the wider representation.  This belief should be tested, and if it
// is false, then we should coalesce these two types into one (the Small width, probably).
// In any case, we believe that is OK to freely cast between these types; no information will
// be lost.
```